src/device/rpi3/timer.go

```diff
@@ -0,0 +1,228 @@
+		h = volatile.LoadUint32((*uint32)(SYSTMR_HI))
+		l = volatile.LoadUint32((*uint32)(SYSTMR_LO))
+	}
+	high := uint64(h << 32)
+	return high | uint64(l)
+}
+
+//Wait n microsec (with BCM System Timer)
+func WaitMuSecST(n uint32) {
+	t := SysTimer()
+	// we must check if it's non-zero, because qemu does not emulate
+	// system timer, and returning constant zero would mean infinite loop
+	if t == 0 {
+		return
+	}
+	for SysTimer() < t+uint64(n) {
+
+	}
+}
+
+//Now returns current time in unix epoch (seconds) format.
+func Now() uint32 {
+	t := SysTimer()
+	t = t - startTicks //how many ticks have we been running
+	t = t / 1000000    //convert to secs
+	return uint32(t) + startTime
+}
+
+//
+// BROADCOM based timer, only works on hardware
+//
+var bcomCurVal uint32 = 0
+
+func BComTimerInit(interval uint32) {
+	bcomCurVal = volatile.LoadUint32((*uint32)(TIMER_CLO))
+	bcomCurVal += interval
+	volatile.StoreUint32((*uint32)(TIMER_C1), bcomCurVal)
+}
+
+func BComHandleTimerIRQ(interval uint32) {
+	bcomCurVal += interval
+	volatile.StoreUint32((*uint32)(TIMER_C1), bcomCurVal)
+	volatile.StoreUint32((*uint32)(TIMER_CS), TIMER_CS_M1)
+	print("IRQ handler: ")
+	UART0Hex(bcomCurVal)
+}
+
+func EnableTimerIRQ() {
+	aarch64.Asm("msr daifclr, #2")
+}
+
+func DisableTimerIRQ() {
+	aarch64.Asm("msr daifset, #2")
+}
+
+func EnableInterruptController() {
+	volatile.StoreUint32((*uint32)(ENABLE_IRQS_1), SYSTEM_TIMER_IRQ_1)
+}
+
+// the exported name is what connects it to the interrupt vector
+//go:export unexpectedInterrupt
+func UnexpectedInterrupt(n int, esr uint64, address uint64) {
+	if n < 0 || n >= len(unexpectedInterruptNames) {
+		print("unexpected interrupt called with bad index", n, "\n")
+		return
+	}
+	print("unexpected interrupt:", unexpectedInterruptNames[n], " esr:", esr, " address 0x")
+	UART0Hex64(address)
+	sp := aarch64.ReadRegister("sp")
+	print("sp is ")
+	UART0Hex64(uint64(sp))
+	aarch64.Asm("eret")
+}
+
+// get the counter frequency
+func CounterFreq() uint32 {
+	var val uint32
+	aarch64.AsmFull(`mrs x28, cntfrq_el0
+		str x28,{val}`, map[string]interface{}{"val": &val})
+	return val
+}
+
+//returns 0 when it doesn't want the timer anymore
+var userCallbackFunc func(uint32, uint64) uint32
+
+// clears the interrupt and writes target value (which is a time distance from now)
+func SetCounterTargetInterval(val uint32, fn func(uint32, uint64) uint32) {
+	aarch64.AsmFull(`mov x28,{val}
+		msr cntv_tval_el0, x28`, map[string]interface{}{"val": val})
+	userCallbackFunc = fn
+}
+
+func CounterTargetVal() uint32 {
+	var val uint32
+	aarch64.AsmFull(`mrs x28,cntv_tval_el0
+		str x28,{val}`, map[string]interface{}{"val": &val})
+	return val
+}
+
+func Core0CounterToCore0Irq() {
+	volatile.StoreUint32((*uint32)(CORE0_TIMER_IRQCNTL), 0x08)
+}
+
+func EnableCounter() {
+	aarch64.Asm(`orr x27, xzr, #1
+		msr cntv_ctl_el0, x27`)
+}
+func UDisableCounter() {
+	aarch64.Asm("msr cntv_ctl_el0, #0")
+}
+
+func WaitForInterrupt() {
+	aarch64.Asm("wfi")
```

Comment:
This function (and a few above: `UDisableCounter`, `EnableCounter`, `Core0CounterToCore0Irq`) arent used anywhere so can be removed.